akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/RotatingKeysSSLEngineProvider.scala

```diff
@@ -0,0 +1,161 @@
+import java.io.IOException
+import java.security.GeneralSecurityException
+import java.security.PrivateKey
+import java.security.SecureRandom
+import java.security.cert.Certificate
+import java.security.cert.X509Certificate
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.actor.ActorSystem
+import akka.annotation.{ ApiMayChange, InternalApi }
+import akka.event.Logging
+import akka.event.MarkerLoggingAdapter
+import akka.remote.artery.tcp.SSLEngineProvider
+import akka.remote.artery.tcp.SecureRandomFactory
+import akka.remote.artery.tcp.SslTransportException
+import akka.remote.artery.tcp.ssl.RotatingKeysSSLEngineProvider.CachedContext
+import akka.remote.artery.tcp.ssl.RotatingKeysSSLEngineProvider.ConfiguredContext
+import akka.stream.TLSRole
+import com.typesafe.config.Config
+import javax.net.ssl.KeyManager
+import javax.net.ssl.SSLContext
+import javax.net.ssl.SSLEngine
+import javax.net.ssl.SSLSession
+import javax.net.ssl.TrustManager
+
+import scala.concurrent.duration._
+
+/**
+ * Variation on ConfigSSLEngineProvider that will periodically reload the keys and certificates
+ * from disk, to facilitate rolling updates of certificates.
+ *
+ * This class is still ApiMayChange because it can likely be further harmonized with
+ * the standard ConfigSSLEngineProvider. Also the location and default values of the
+ * configuration may change in future versions of Akka.
+ *
+ * This provider does not perform hostname verification, but instead allows checking
+ * that the remote certificate has a subject name that matches the subject name of
+ * the configured certificate.
+ */
+@ApiMayChange
+final class RotatingKeysSSLEngineProvider(val config: Config, protected val log: MarkerLoggingAdapter)
+    extends SSLEngineProvider {
+
+  def this(system: ActorSystem) =
+    this(
+      system.settings.config.getConfig("akka.remote.artery.ssl.rotating-keys-engine"),
+      Logging.withMarker(system, classOf[RotatingKeysSSLEngineProvider].getName))
+
+  // read config
+
+  private val SSLKeyFile: String = config.getString("key-file")
+  private val SSLCertFile: String = config.getString("cert-file")
+  private val SSLCACertFile: String = config.getString("ca-cert-file")
+
+  private val sslEngineConfig = new SSLEngineConfig(config)
+  import sslEngineConfig._
+
+  // build a PRNG (created once, reused on every instance of SSLContext
+  private val rng: SecureRandom = SecureRandomFactory.createSecureRandom(SSLRandomNumberGenerator, log)
+
+  // handle caching
+  private val contextRef = new AtomicReference[Option[CachedContext]](None)
+
+  /** INTERNAL API */
+  @InternalApi
+  private[ssl] def getSSLContext() = getContext().context
+  private def getContext(): ConfiguredContext = {
+    contextRef.get() match {
+      case Some(CachedContext(_, expired)) if expired.isOverdue() =>
+        val context = constructContext()
+        contextRef.set(Some(CachedContext(context, SSLContextCacheTime.fromNow)))
```

Comment:
maybe add a comment that it's fine to create new context in case of concurrent access